### PR TITLE
CPM-558: Add query to fetch non viewable products in list

### PR DESCRIPTION
**Pull request by Balamung**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/EnrichmentProductTestCase.php [hunk 1]
```diff
@@ -213,4 +213,14 @@ protected function clearDoctrineUoW(): void
     {
         $this->get('pim_connector.doctrine.cache_clearer')->clear();
     }
+
+    /**
+     * @return array<string>
+     */
+    protected function getAssociatedProductIdentifiers(ProductInterface $product): array
```
Comment (cemorin) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/EnrichmentProductTestCase.php:220]: its already in other 258?

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php [hunk 2]
```diff
@@ -59,8 +61,9 @@ public function apply(UserIntent $userIntent, ProductInterface $product, int $us
                     continue;
                 }
 
+                $nonViewableProducts = $this->getNonViewableProducts->fromProductIdentifiers($formerAssociations, $userId);
                 $normalizedAssociations[$associationUserIntent->associationType()][$entityType] = \array_values(
-                    \array_unique($associationUserIntent->productIdentifiers())
+                    \array_merge($nonViewableProducts, \array_unique($associationUserIntent->productIdentifiers()))
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:66]:
```suggestion
                    \array_unique(\array_merge($nonViewableProducts, $associationUserIntent->productIdentifiers()))
```

it should be the same, but it's safer like this IMO

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Query/GetNonViewableProducts.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Akeneo\Pim\Enrichment\Product\Infrastructure\Query;
+
+use Akeneo\Pim\Enrichment\Product\Domain\Query\GetNonViewableProducts as GetNonViewableProductsInterface;
+
+/**
+ * @copyright 2022 Akeneo SAS (https://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class GetNonViewableProducts implements GetNonViewableProductsInterface
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Query/GetNonViewableProducts.php:13]:
```suggestion
final class GetNonViewableProducts implements GetNonViewableProductsInterface
```

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php [hunk 4]
```diff
@@ -208,4 +213,53 @@ public function it_merges_non_viewable_category_on_update(): void
...
+        $this->messageBus->dispatch($command);
+
+        $this->getContainer()->get('pim_catalog.validator.unique_value_set')->reset();
+        $this->clearDoctrineUoW();
+
+        $product = $this->productRepository->findOneByIdentifier('my_product');
+
+        Assert::assertNotNull($product);
+        Assert::assertSame('my_product', $product->getIdentifier());
+        Assert::assertEqualsCanonicalizing(['my_former_product', 'my_former_product_2', 'my_other_product'], $this->getAssociatedProductIdentifiers($product));
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php:263]: Could you rename `my_former_product_2` into something like `product_non_viewable_by_redactor` to ease the understanding?  

And I'm not sure to understand why `my_former_product` is still associated? This product has no category so it should be viewable

**Review by nmarniesse — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php (hunk 4 above)
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php]: Yes the tests wasn't passing, my bad (there is no `my_former_product` anymore)

**Review by ahocquard — COMMENTED** [review]
Feedback after merged

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php [hunk 5]
```diff
@@ -22,6 +23,7 @@ final class AssociationUserIntentCollectionApplier implements UserIntentApplier
 {
     public function __construct(
         private ObjectUpdaterInterface $productUpdater,
+        private GetNonViewableProducts $getNonViewableProducts
```
[the remark this answers is not in the document]
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:26]: This PR has been merged into another one here: https://github.com/akeneo/pim-community-dev/pull/16711 so I can still make the changes :+1:

**Pull request merged**